os/drivers/input/ist415.c

```diff
@@ -89,185 +98,1910 @@ struct touchscreen_ops_s g_ist415_ops = {
+	uint8_t data;
+
+	if (dev->rec_mode) {
+		data = REC_ENABLE;
+		ret = ist415_i2c_write(dev, HCOM_SET_REC_MODE, &data, 1);
+		if (ret) {
+			touchdbg("ERROR: fail to write Recording mode command (Enable)\n");
+		} else {
+			touchvdbg("Recording mode start\n");
+		}
+	}
+
+	ist415_touchtype(dev, (uint8_t *)&dev->touchtype);
+	ist415_power_mode(dev, dev->sysmode);
+	ist415_sensor(dev, true);
+
+	if (dev->rec_mode) {
+		// TODO: Delay 100ms
+		usleep(100000);
+		data = REC_START_SCAN;
+		ret = ist415_i2c_write(dev, HCOM_SET_REC_MODE, &data, 1);
+		if (ret) {
+			touchdbg("ERROR: fail to write Recording mode command (Start scan)\n");
+		}
 	}
+
+	touchvdbg("ist415 start\n");
 }
 
 /****************************************************************************
- * Public Functions
+ * Name: ist415_isp_enable
  ****************************************************************************/
 
+static int ist415_isp_enable(struct ist415_dev_s *dev, bool enable)
+{
+	int ret = 0;
+
+	if (enable) {
+		ret = ist415_i2c_write_isp(dev, ISP_TARGET_ENTER, ISP_CMD_ENTER, 0, NULL, 0);
+	} else {
+		ret = ist415_i2c_write_isp(dev, ISP_TARGET_EXIT, ISP_CMD_EXIT, 0, NULL, 0);
+	}
+
+	return ret;
+}
+
 /****************************************************************************
- * Name: ist415_touchinitialize
- *
- * Description:
- *   This function is called by board-specific, setup logic to configure
- *   and register the IST415 Touch device.  This function will register the driver
- *   as /dev/touchN where N is determined by the minor device number.
- *
- * Input Parameters:
- *   i2c_dev - The i2c device structure for the I2C bus that the touch controller is connected to
- *   config - The lower half touch driver structure that contains the touch configuration information.
- *
- * Returned Value:
- *   On success, a non-NULL pointer to the touch device structure touch_dev_s is returned.
- *   On a failure, NULL is returned.
- *
+ * Name: ist415_isp_erase
  ****************************************************************************/
 
-int ist415_initialize(const char*path, struct ist415_dev_s *priv)
+static int ist415_isp_erase(struct ist415_dev_s *dev, uint8_t target, uint8_t area, uint32_t addr, int size)
 {
 	int ret = 0;
-	uint8_t reg[1];
-	priv->ops->irq_disable();
+	int count = 0;
+	int unit = 0;
 
-	reg[0] = 0x23;
-	ret = i2c_write(priv->i2c, &(priv->i2c_config), (uint8_t *)reg, 1);
-	if (ret < OK) {
-		touchdbg("ERROR: i2c_write failed\n");
-		return NULL;
+	touchvdbg("%s\n", __func__);
+
+	if (area == ISP_CMD_MASS_ERASE) {
+		count = 1;
+	} else if (area == ISP_CMD_SECTOR_ERASE) {
+		count = size / IST415_ROM_SECTOR_SIZE;
+		unit = IST415_ROM_SECTOR_SIZE;
+	} else if (area == ISP_CMD_PAGE_ERASE) {
+		count = size / IST415_ROM_PAGE_SIZE;
+		unit = IST415_ROM_PAGE_SIZE;
+	}
+
+	for (int i = 0; i < count; i++) {
+		ret = ist415_i2c_write_isp(dev, target, area, addr + (i * unit), NULL, 0);
+		if (ret < 0) {
+			touchdbg("ERROR: ISP erase fail\n");
```

Comment:
should it continue the for loop if ISP erase is failed. I think it should return from here